diff -urN --exclude po logjam-4.2.4/src/logjam-doc.c logjam-4.2.4-my/src/logjam-doc.c
--- logjam-4.2.4/src/logjam-doc.c	Sat Jun  7 02:02:21 2003
+++ logjam-4.2.4-my/src/logjam-doc.c	Tue Sep  9 19:45:39 2003
@@ -401,6 +401,50 @@
 }
 
 gboolean
+logjam_doc_insert_popen(LogJamDoc *doc, const char *execstring) {
+	gchar buf[BUFSIZ];
+	GString * output;
+	const gchar *end;
+	int ret;
+	FILE* fl;
+
+	fl = popen(execstring, "r");
+	if (fl == NULL) {
+		/* FIXME: show error */
+		return FALSE;
+	}
+
+	output = g_string_sized_new (sizeof(buf));
+	while (fgets(buf, sizeof(buf), fl) != NULL) {
+		g_string_append(output, buf);
+	}
+	ret = pclose(fl);
+	if (ret != 0) {
+		/* FIXME: show error */
+		return FALSE;
+	}
+
+	if (g_utf8_validate(output->str, output->len, &end)) {
+		gtk_text_buffer_insert_at_cursor(doc->buffer,
+				output->str, output->len);
+	} else {
+		gchar* newtext;
+
+		newtext = g_locale_to_utf8 (output->str, output->len,
+				NULL, NULL, NULL);
+		if (!newtext) {
+			/* FIXME: show error */
+			g_string_free(output, TRUE);
+			return FALSE;
+		}
+		gtk_text_buffer_insert_at_cursor(doc->buffer, newtext, -1);
+		g_free(newtext);
+	}
+	g_string_free(output, TRUE);
+	return TRUE;
+}
+
+gboolean
 logjam_doc_insert_file(LogJamDoc *doc, const char *filename, const char *encoding, GError **err) {
 	char *text;
 	gsize len;
diff -urN --exclude po logjam-4.2.4/src/logjam-doc.h logjam-4.2.4-my/src/logjam-doc.h
--- logjam-4.2.4/src/logjam-doc.h	Sat Jun  7 02:02:22 2003
+++ logjam-4.2.4-my/src/logjam-doc.h	Tue Sep  9 17:49:23 2003
@@ -64,6 +64,7 @@
 
 
 gboolean logjam_doc_insert_file(LogJamDoc *doc, const char *filename, const char *encoding, GError **err);
+gboolean logjam_doc_insert_popen(LogJamDoc *doc, const char *execstring);
 
 EntryType logjam_doc_get_entry_type(LogJamDoc *doc);
 gint logjam_doc_get_entry_itemid(LogJamDoc *doc);
diff -urN --exclude po logjam-4.2.4/src/menu.c logjam-4.2.4-my/src/menu.c
--- logjam-4.2.4/src/menu.c	Mon Jun  2 07:16:08 2003
+++ logjam-4.2.4-my/src/menu.c	Tue Sep  9 17:49:23 2003
@@ -256,6 +256,12 @@
 menu_insert_file(LJWin *ljw) {
 	tools_insert_file(GTK_WINDOW(ljw), ljw->doc);
 }
+
+static void
+menu_insert_popen(LJWin *ljw) {
+	tools_insert_popen(GTK_WINDOW(ljw), ljw->doc);
+}
+
 static void
 menu_html_escape(LJWin *ljw) {
 	tools_html_escape(GTK_WINDOW(ljw), ljw->doc);
@@ -354,6 +360,7 @@
 { N_("/_Insert"),                 NULL, NULL, 0, "<Branch>" },
 { N_("/Insert/_Poll..."),  NULL, run_poll_creator_dlg, 0, NULL },
 { N_("/Insert/_File..."),  NULL, menu_insert_file, 0, NULL },
+{ N_("/Insert/_Output of command..."),  NULL, menu_insert_popen, 0, NULL },
 {    "/Insert/---",               NULL, NULL, 0, "<Separator>" },
 { N_("/Insert/_Link..."),         "<ctl>L",       menu_make_link },
 { N_("/Insert/_Journal Link..."), "<ctl><alt>L", menu_make_journal_link },
diff -urN --exclude po logjam-4.2.4/src/tools.c logjam-4.2.4-my/src/tools.c
--- logjam-4.2.4/src/tools.c	Sat Jun  7 02:02:22 2003
+++ logjam-4.2.4-my/src/tools.c	Tue Sep  9 19:52:43 2003
@@ -138,6 +138,24 @@
 }
 
 void
+tools_insert_popen(GtkWindow *win, LogJamDoc *doc) {
+	char* execcommand;
+
+	execcommand = lj_command_get(win, NULL);
+	if (execcommand) {
+		/* we'll be blocked for some time (while excuting popen()),
+		 * so process all events before blocking */
+		gdk_window_process_all_updates();
+		if (!logjam_doc_insert_popen(doc, execcommand)) {
+			lj_warning(win,	_("Error executing command %s"),
+				execcommand);
+		}
+		g_free(execcommand);
+	}
+}
+
+
+void
 tools_insert_file(GtkWindow *win, LogJamDoc *doc) {
 	GtkWidget *filesel;
 	GtkWidget *hbox, *label, *combo;
diff -urN --exclude po logjam-4.2.4/src/tools.h logjam-4.2.4-my/src/tools.h
--- logjam-4.2.4/src/tools.h	Mon Jun  2 07:16:08 2003
+++ logjam-4.2.4-my/src/tools.h	Tue Sep  9 17:49:23 2003
@@ -12,6 +12,7 @@
 void tools_html_escape      (GtkWindow *win, LogJamDoc *doc);
 void tools_remove_linebreaks(GtkWindow *win, LogJamDoc *doc);
 void tools_insert_file      (GtkWindow *win, LogJamDoc *doc);
+void tools_insert_popen     (GtkWindow *win, LogJamDoc *doc);
 void tools_validate_xml     (GtkWindow *win, LogJamDoc *doc);
 void tools_ljcut            (GtkWindow *win, LogJamDoc *doc);
 
diff -urN --exclude po logjam-4.2.4/src/util.c logjam-4.2.4-my/src/util.c
--- logjam-4.2.4/src/util.c	Sun Jun 15 21:37:23 2003
+++ logjam-4.2.4-my/src/util.c	Tue Sep  9 19:11:55 2003
@@ -664,6 +664,37 @@
 }
 
 char*
+lj_command_get(GtkWindow *parent, const char *basename) {
+	GtkWidget *dlg, *box, *entry;
+	
+	dlg = gtk_dialog_new_with_buttons(_("Insert command output"),
+			parent, GTK_DIALOG_MODAL,
+			GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
+			GTK_STOCK_OK, GTK_RESPONSE_OK,
+			NULL);
+	gtk_window_set_transient_for(GTK_WINDOW(dlg), parent);
+	gtk_dialog_set_default_response(GTK_DIALOG(dlg), GTK_RESPONSE_OK);
+
+	entry = gtk_entry_new();
+	if (basename)
+		gtk_entry_set_text(GTK_ENTRY(entry), basename);
+	gtk_entry_set_activates_default(GTK_ENTRY(entry), TRUE);
+
+	box = labelled_box_new(_("_Command:"), entry);
+	lj_dialog_set_contents(GTK_DIALOG(dlg), box);
+	gtk_widget_grab_focus(entry);
+
+	if (gtk_dialog_run(GTK_DIALOG(dlg)) == GTK_RESPONSE_OK) {
+		char * text;
+		text = g_strdup (gtk_entry_get_text(GTK_ENTRY(entry)));
+		gtk_widget_destroy(dlg);
+		return text;
+	}
+	gtk_widget_destroy(dlg);
+	return NULL;
+}
+
+char*
 get_summary(char *subject, char *event) {
 #define SUMMARY_LENGTH 50
 	/* SUMMARY_LENGTH is in chars, not bytes.
diff -urN --exclude po logjam-4.2.4/src/util.h logjam-4.2.4-my/src/util.h
--- logjam-4.2.4/src/util.h	Sun Jun 15 21:37:23 2003
+++ logjam-4.2.4-my/src/util.h	Tue Sep  9 17:49:23 2003
@@ -50,6 +50,7 @@
 void lj_win_set_size(GtkWindow *win, int width, int height);
 
 GtkWidget* lj_filesel_new(const gchar *title, GtkWindow *parent);
+char* lj_command_get(GtkWindow *parent, const char *basename);
 
 void lj_dialog_set_contents_container(GtkDialog *dlg, GtkWidget *container);
 GtkWidget* lj_dialog_set_contents(GtkDialog *dlg, GtkWidget *contents);
